torch/distributed/_composable/fsdp/_fsdp_param.py

```diff
@@ -93,6 +94,18 @@ class ParamModuleInfo:
     shared_param_names: List[str] = field(default_factory=list)
 
 
+@dataclass
+class ExtensionsData:
+    # User-defined metadata passed from pre to post-all-gather
+    all_gather_metadata: Optional[Any] = None
+    # Save the all-gather input sizes to unflatten the all-gather outputs to ND
+    all_gather_input_sizes: Sequence[torch.Size] = ()  # ND
```

Comment:
wondering this `ExtensionData` can be merged together with the default FSDPParam metas?

Looks like simple `torch.Tensor` case is just one case of this

Reply:
I think one thing is that the `ExtensionsData` will never be used for the simple `torch.Tensor` case:
- The `all_gather_metadata` is specific to the extensions so that the user can pass metadata between the pre and post-all-gather.
- The `all_gather_input_sizes` are not needed since we do not need to unflatten the all-gather outputs to ND (but rather just keep as 1D).